### PR TITLE
Authenticated git repository access

### PR DESCRIPTION
**Pull request by giuliojiang**
You can now provide username&password for a gitlab account that has access to student & model repositories.

### CODE REVIEW

File: infrastructure/gitauth.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        return
+    
+    username = ''
+    password = ''
+    print('Please insert your gitlab authentication details. These will be used to clone the student repositories and model answers. If they are left blank, only publicly accessible repositories can be cloned.')
+    print('Username:')
+    username = raw_input()
+    # Need to use URL encoding
+    username = urllib.quote(username)
+    password = getpass.getpass()
```
Comment (AbraaoMota) [inline infrastructure/gitauth.py:32]: How does this know the password of the user? Also, where is this shown to the user? When he starts the server?

**Review by AbraaoMota — APPROVED** [review]

File: infrastructure/gitauth.py (hunk 1 above)
Comment (giuliojiang) [reply infrastructure/gitauth.py:32]: this asks the user to input a password when RunServer is run (similarly to how we ask for the db password)

Comment (giuliojiang) [reply infrastructure/gitauth.py:32]: getpass() is like raw_input() but it doesn't print it to stdout (no tee)

**Review by procaconsul — APPROVED** [review]

**Comment by helicopter88** [comment]
Seems okay, I am worried by the fact that passwords are stored in plain text.
Also, why cannot I use git through the ssh url? it is much safer

**Comment by giuliojiang** [comment]
git authentication data is not plaintext anymore

**Pull request merged**